src/main/kotlin/org/rust/ide/template/postfix/utils.kt

```diff
@@ -0,0 +1,66 @@
+package org.rust.ide.template.postfix
+
+import com.intellij.codeInsight.template.postfix.templates.PostfixTemplateExpressionSelector
+import com.intellij.codeInsight.template.postfix.templates.PostfixTemplatePsiInfo
+import com.intellij.openapi.editor.Document
+import com.intellij.psi.PsiElement
+import com.intellij.util.Function
+import org.rust.lang.core.psi.RustBlockElement
+import org.rust.lang.core.psi.RustElementFactory
+import org.rust.lang.core.psi.RustExprElement
+import org.rust.lang.core.psi.util.ancestors
+import org.rust.lang.core.types.RustBooleanType
+import org.rust.lang.core.types.util.resolvedType
+import org.rust.lang.utils.negate
+
+internal object RustPostfixTemplatePsiInfo : PostfixTemplatePsiInfo() {
+    override fun getNegatedExpression(element: PsiElement): PsiElement =
+        element.negate()
+
+    override fun createExpression(context: PsiElement, prefix: String, suffix: String): PsiElement =
+        RustElementFactory.createExpression(context.project, "$prefix${context.text}$suffix")!!
+}
+
+abstract class RustExprParentsSelectorBase(val pred: (RustExprElement) -> Boolean) : PostfixTemplateExpressionSelector {
+    override fun getRenderer(): Function<PsiElement, String> = Function {
+        e: PsiElement ->
+        e.text
+    }
+
+    abstract override fun getExpressions(context: PsiElement, document: Document, offset: Int): MutableList<PsiElement>
+}
+
+class RustAllParentsSelector(pred: (RustExprElement) -> Boolean) : RustExprParentsSelectorBase(pred) {
+    override fun hasExpression(context: PsiElement, document: Document, offset: Int): Boolean =
+        context
+            .ancestors
+            .filter { it is RustExprElement && pred(it) }
+            .any()
+
+    override fun getExpressions(context: PsiElement, document: Document, offset: Int): MutableList<PsiElement> =
+        context
+            .ancestors
+            .filter { it is RustExprElement && pred(it) }
+            .toMutableList()
+}
+
+class RustTopMostInScopeSelector(pred: (RustExprElement) -> Boolean) : RustExprParentsSelectorBase(pred) {
+    override fun getExpressions(context: PsiElement, document: Document, offset: Int): MutableList<PsiElement> =
+        mutableListOf(
+            context
+                .ancestors
+                .takeWhile { it !is RustBlockElement }
+                .filter { it is RustExprElement && pred(it) }
+                .last()
+        )
+
+    override fun hasExpression(context: PsiElement, copyDocument: Document, newOffset: Int): Boolean =
+        context
+            .ancestors
+            .takeWhile { it !is RustBlockElement }
+            .filter { it is RustExprElement && pred(it) }
+            .lastOrNull() !=
+            null
+}
+
+fun RustExprElement.IS_BOOL() = resolvedType == RustBooleanType
```

Comment:
This function is used only in `surrounderBasedPostfixTemplates`, so let's move it there, make it private and use a more :camel:ventional name `isBool`.